File: vainilla/src/main/java/extensions/movements/Movable.java
package extensions.movements;


import com.uqbar.vainilla.DeltaState;
import com.uqbar.vainilla.UnitVector2D;


/**
 * Defines a movable object. It can be moved by a deltaState and a speed.
 */
public interface Movable {

	public void move(DeltaState deltaState);

	public double move(double position, double uVectorProportion, double speedFactor);

	public double getSpeedFactor(DeltaState deltaState);

	public double getSpeed();

	public UnitVector2D getUVector();

	public double getX();

	public double getY();

}

File: pacman-reloaded/src/main/java/components/mixins/movable.rb
require 'java'

java_import Java::com.uqbar.vainilla.UnitVector2D

module Movable
  attr_accessor :uVector, :speed

  def initialize_vector(xVec, yVec, speed)
    self.uVector = UnitVector2D.new(xVec, yVec)
    self.speed = speed
  end

  def move(delta_state)
    self.x = move_position x, uVector.x, speed_factor(delta_state)
    self.y = move_position y, uVector.y, speed_factor(delta_state)
  end

  def move_position(position, uVectorProportion, speedFactor)
    position + uVectorProportion * speedFactor
  end

  def speed_factor(delta_state)
    speed * delta_state.delta
  end
end